Add GOI analysis calculations script

cbethell opened the pull request:
**Issue Addressed**
Closes #172 

**What is the purpose of these changes?**
<!--Provide some background on the changes proposed.-->
This PR initiations a `goi-calculations.R` script in the optional-goi-analysis module, and modifies the relevant functions script appropriately.

**What changes did you make?**
<!--Describe the concrete changes that you made or additional features that were added, be as detailed as possible in the steps you took.-->
As noted on issue #172, this PR moves the optional gene identifier mapping from a separate R script to the first step of the GOI analysis module which is the `goi-calculations.R` script.

Within the calculations script is also the preparation of the heatmap matrix and annotation, thus removing the need for any calculations to be done in the goi template notebook.

I have also added a short example GOI list to `example-data/goi-lists` for testing purposes (suggestions for optimization here are welcome)!

There is some added logic here as well, to account for the need for Ensembl ids to filter the matrix and run the `colnames_to_gene_symbols()` function in the script, please let me know if you have any suggestions here we well. See below for reference to the code being referred to here:

<img width="685" alt="Screen Shot 2022-10-20 at 2 16 11 PM" src="https://user-images.githubusercontent.com/43576623/197038200-29caa7e9-d04c-4784-bd28-ebb20a082652.png">

**Any comments, concerns, or questions important for reviewers**
- Does this PR seem to adequately address #172? 
- Are there any suggestions for refactoring?

Here is also a screenshot of the output structure currently within one sample's results folder:

<img width="309" alt="Screen Shot 2022-10-20 at 2 19 01 PM" src="https://user-images.githubusercontent.com/43576623/197038636-32265d27-1078-42e1-a86d-bbb2a2d5c59f.png">

**Checklist**

Place an `x` in all boxes that you have completed.

- [x] I have run the most recent version of the code
- [ ] If I am adding in reports or generating any plots, I have reviewed the necessary reports
- [ ] I have added all necessary documentation (if applicable)

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    c("-l", "--library_id"),
+    type = "character",
+    default = NULL,
+    help = "name of library being analyzed",
```
```suggestion
    help = "Unique ID corresponding to library being analyzed",
```
Just want to make sure we specify that each library ID should be unique.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    c("-s", "--seed"),
+    type = "integer",
+    default = 2021,
+    help = "seed integer",
```
```suggestion
    help = "Random seed to set",
```

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    c("-g", "--input_goi_list"),
+    type = "character",
+    default = NULL,
+    help = "file path to the unmapped input genes of interest list"
```
```suggestion
    help = "file path to the input genes of interest list"
```
Doesn't necessarily need to be unmapped

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    default = NULL,
+    help = "type of gene identifiers to be mapped to and returned"
+  ),
+  optparse::make_option(
+    c("-c", "--identifier_column_name"),
+    type = "character",
+    default = NULL,
+    help = "the name of the column in the input file that contains the gene
+            identifiers to be converted from"
+  ),
```
I think we might actually be able to remove this and specify in the docs what the column name should be? We could use a generic `gene_id` for the column name for the input rather than having people set their own column name.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    help = "the name of the column in the input file that contains the gene
+            identifiers to be converted from"
+  ),
+  optparse::make_option(
+    c("--gene_set_column_name"),
+    type = "character",
+    default = NULL,
+    help = "the name of the column in the input file that contains the gene set
+            information"
+  ),
```
This is an additional grouping right? Like if a gene belongs to a group of genes? We should maybe expand our example file to include a gene set column. Again, I would also set a required name for this column like `gene_set` and make that column optional.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+    help = "the scientific name of the organism relevant to the genes to be
+            mapped, 'Homo sapiens' or 'Mus musculus' for examples"
+  ),
+  optparse::make_option(
+    c("-m", "--multi_mappings"),
+    type = "character",
+    default = NULL,
+    help = "how should instances of multiple gene identifier mappings be handled
+            - may specify 'list' to return all the mappings or 'first' to keep
+            only the first mapping"
+  ),
```
I think here you want to set a default of what we would recommend, which is probably `first`. Also if someone were to use `list` how would that be incorporated into the downstream plots? We probably want to always include a column with the first mapping and then if someone uses `list` then they would receive an additional column with secondary mappings, but the plots will only include the first mapping.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+#### Read in data --------------------------------------------------------------
+
+goi_list <- data.table::fread(opt$input_goi_list, stringsAsFactors = FALSE)
+sce <- read_rds(file.path(opt$sce))
```
This is where you would want to add a check to make sure that the SCE object is in the correct format, e.g. contains normalized data

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+
+#### Prepare data for plotting -------------------------------------------------
+
+identifier_column_name_sym <- rlang::sym(opt$identifier_column_name)
```
You don't actually use this anywhere else? I think you could delete it?

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+#### Perform mapping -----------------------------------------------------------
+
+if (opt$perform_mapping == TRUE) {
+  # turn column name into symbol for pulling the column info out of data frame
```
Before performing any mapping you probably want a check that provided key types are present in the annotations.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+           'ENSEMBL' for example, see `keytypes()` in the organism's annotation
+            package for more options"
+  ),
+  optparse::make_option(
+    c("--output_identifiers"),
+    type = "character",
+    default = NULL,
+    help = "type of gene identifiers to be mapped to and returned"
+  ),
```
```suggestion
  optparse::make_option(
    c("--sce_rownames_identifier"),
    type = "character",
    default ="ENSEMBL",
    help = "gene identifier type associated with rownames of provided SCE object"
  ),
```
In thinking about how we want to frame this, I think we want to refer to this as the keytype associated with the rownames in the SCE object. The goal of this is to get the input genes in the same format as those row names, which by default in our pipeline will be ensembl gene IDs. You would need to propagate this variable name change throughout.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+  write_tsv(goi_list, file.path(
+    opt$output_directory,
+    paste0(opt$library_id, "_mapped_genes.tsv")
+  ))
+}
```
```suggestion
  ))
  
  # define mapped goi associated with rownmaes of SCE object 
  goi_rownames_column <- tolower(opt$sce_rownames_identifier)
  goi_rownames <- goi_list %>%
    dplyr::pull(goi_rownames_column)
    
} else {

  # if no mapping is performed then set goi to input id column 
  goi_rownames <- goi_list %>%
    dplyr::pull(gene_id)
  }
```

Here I would set the variables for grabbing the genes from the SCE object based on if mapping was performed or not. I also used my suggestion from earlier on setting the input ID column to always be `gene_id` and removing that argument entirely.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,301 @@
+
+# filter counts matrix to only data associated with the provided genes of
+# interest
+normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% goi_list[[ensembl_id_column]]]
```
```suggestion
normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% goi_rownames]
```
simplifying a little with my previous suggestion

allyhawkins reviewed (commented):
Thanks for getting this started! I have some comments throughout to help narrow down the arguments needed and refactor a little/ add in some checks. I probably have more comments on how to clean up preparation of the heatmap, but want to get these to you right now and then we can work through that part.

cbethell commented:
Thank you for the review @allyhawkins, it has been extremely helpful!

I was able to implement your refactoring suggestions, add the example `gene_set` column to the goi list added here, and include the necessary checks. 

The only comments I want to shed light on at this time are:
- > I think here you want to set a default of what we would recommend, which is probably first. Also if someone were to use list how would that be incorporated into the downstream plots? We probably want to always include a column with the first mapping and then if someone uses list then they would receive an additional column with secondary mappings, but the plots will only include the first mapping.
https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/230#discussion_r1004640593

As we discussed yesterday, I agree with you here but have not pinned down how to ideally tackle this as yet (with the lack of example multi-mapping data). Let me know if you have anything readily in mind, otherwise I will dig further into this. Also let me know if it would be easier to you to pair on this!

- > This is where you would want to add a check to make sure that the SCE object is in the correct format, e.g. contains normalized data
https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/230#discussion_r1004642646

I added a check that `logcounts(sce)` is not NULL but please let me know if you had more in mind regarding what should be checked for at this point!

Overall let me know what you think of things now!

cbethell commented:
Committed the further refactoring changes we discussed in Science Pair It @allyhawkins, _now_ I believe this is ready for another look!

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+#   --input_goi_list "example-data/goi-lists/sample01_goi_list.tsv" \
+#   --perform_mapping TRUE \
+#   --sce_rownames_identifier "ENSEMBL" \
+#   --provided_identifier <- "SYMBOL" \
```
```suggestion
#   --provided_identifier "SYMBOL" \
```

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+  ),
+  optparse::make_option(
+    c( "--perform_mapping"),
+    action = "store_false",
+    help = "specifies whether or not to perform gene identifier mapping; the 
+    default is FALSE as mapping is not required when ENSEMBL ids are provided"
```
```suggestion
    action = "store_true",
    default = FALSE,
    help = "specifies whether or not to perform gene identifier mapping; the 
    default is FALSE as mapping is not required when ENSEMBL ids are provided"
```
Based on the description I think what you mean here is to use `store_true` which will set this to true by using the `--perform_mapping` flag at the command line and then it will be false if the flag is not used.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+#   --library_id "library01" \
+#   --seed 2021 \
+#   --input_goi_list "example-data/goi-lists/sample01_goi_list.tsv" \
+#   --perform_mapping TRUE \
```
```suggestion
#   --perform_mapping \
```
Shouldn't need to specify true/false with `action=store_true`

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+    c("-g", "--input_goi_list"),
+    type = "character",
+    default = NULL,
+    help = "file path to the input genes of interest list"
```
```suggestion
    help = "file path to the input genes of interest list, must contain gene identifiers to plot in a column named `gene_id`."
```

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+#   --sce_rownames_identifier "ENSEMBL" \
+#   --provided_identifier <- "SYMBOL" \
+#   --organism "Homo sapiens" \
+#   --multi_mappings "list" \
```
```suggestion
#   --multi_mappings "first" \
```

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+    c("--output_directory"),
+    type = "character",
+    default = NULL,
+    help = "path to output directory that would hold the TSV files with the 
+    clustering stats"
```
```suggestion
    help = "path to output directory"
```

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+if(!(any(c("PCA", "UMAP") %in% reducedDimNames(sce)))){
+  stop("There are no PCA or UMAP results in the provided SingleCellExperiment object.
+       You may want to run the core workflow analysis before re-running this GOI analysis.")
+}
```
One other check I thought of we are missing is that `gene_id` is a column name in the input GOI list.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+#### Prepare data for plotting -------------------------------------------------
+
+# get logcounts from normalized sce object
+normalized_sce_logcounts_matrix <- as.matrix(t(logcounts(sce)))
```
```suggestion
normalized_logcounts_matrix <- as.matrix(t(logcounts(sce)))
```
I think throughout you could remove the `_sce_` from the variable names since these are now matrices and not SCE objects.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,291 @@
+}
+
+# Save matrix object to file
+write_rds(normalized_sce_zscores_matrix,
+          file.path(
+            opt$output_directory,
+            paste0(opt$library_id, "_normalized_sce_zscores.rds")
+          ))
```
This is a matrix correct? We could save this as a compressed `.mtx` file instead of an Rds file?

allyhawkins reviewed (commented):
Just a few more minor comments, but the logic of everything looks good now!

cbethell commented:
Thanks @allyhawkins, I've implemented your additional suggestions with exception of saving the matrix as a `.mtx` file. I tried to do this using the `Matrix::writeMM()` function but received the following error after confirming the package is indeed installed, any ideas?

<img width="616" alt="Screen Shot 2022-11-02 at 5 15 51 PM" src="https://user-images.githubusercontent.com/43576623/199613255-30d6e7e8-156f-4d60-83b0-87d3e9411663.png">

allyhawkins commented:
> Thanks @allyhawkins, I've implemented your additional suggestions with exception of saving the matrix as a .mtx file. I tried to do this using the Matrix::writeMM() function but received the following error after confirming the package is indeed installed, any ideas?

This is because it looks like the scaled logcounts matrix that you're trying to write out is actually a dense matrix and not a sparse matrix, which is what `writeMM` requires. I would first convert to a sparse matrix and then write it out.

cbethell commented:
> This is because it looks like the scaled logcounts matrix that you're trying to write out is actually a dense matrix and not a sparse matrix, which is what `writeMM` requires. I would first convert to a sparse matrix and then write it out.

Ah yes, that worked and makes sense why! Thank you @allyhawkins!

allyhawkins reviewed (approved):
LGTM 🚀

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,303 @@
+    dplyr::distinct() %>%
+    dplyr::left_join(goi_list, by = "gene_id")
+  
+  # Save mapped object to file
+  write_tsv(goi_list, file.path(
+    opt$output_directory,
+    paste0(opt$library_id, "_mapped_genes.tsv")
+  ))
```
```suggestion
```

As I started to review #232, I realized that we are always taking the mapped goi list as input, but if mapping is false that file won't exist. So what we need to do is move writing out the file to be outside of the if statement and write out a new goi list that gets stored in the goi stats directory regardless. I think we could still name this as with the `_mapped_genes.tsv` extension, but it will only contain the additional mapped columns if mapping was performed, otherwise it will be the same as the input goi list.

allyhawkins commented on optional-goi-analysis/goi-calculations.R:
```diff
@@ -0,0 +1,303 @@
+  goi_rownames <- goi_list %>%
+    dplyr::pull(gene_id)
+}
+
```
```suggestion
 # Save goi list to file to be used as input for template 
  write_tsv(goi_list, file.path(
    opt$output_directory,
    paste0(opt$library_id, "_mapped_genes.tsv")
  ))
```

allyhawkins reviewed (commented):
Sorry one more change I found we need.

cbethell commented:
I have fixed the way we save the output mapped file per https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/230#discussion_r1014215860 and also added the `sce_rownames_identifier` column here per https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/232#discussion_r1014238849. I _believe_ this should be ready now!

allyhawkins commented:
Took a quick glance and the changes look correct! Should be good to go to merge.
